addons/website_sale/controllers/main.py

```diff
@@ -1110,6 +1133,150 @@ def address(self, **kw):
+        elif any(billing_address[k] != order_sudo.partner_invoice_id[k] for k in billing_address):
+            # Check if a child partner doesn't already exist with the same informations. The
+            # phone isn't always checked because it isn't sent in shipping information with
+            # Google Pay.
+            child_partner_id = self._find_child_partner(
+                order_sudo.partner_id.commercial_partner_id.id, billing_address
+            )
+            if child_partner_id:
+                order_sudo.partner_invoice_id = child_partner_id
+            else:
+                billing_partner_id = self._create_or_edit_partner(
+                    billing_address,
+                    type='invoice',
+                    parent_id=order_sudo.partner_id.id,
+                )
+                order_sudo.partner_invoice_id = billing_partner_id
+
+        # In a non-express flow, `sale_last_order_id` would be added in the session before the
+        # payment. As we skip all the steps with the express checkout, `sale_last_order_id` must be
+        # assigned to ensure the right behavior from `shop_payment_confirmation()`.
+        request.session['sale_last_order_id'] = order_sudo.id
+
+        return order_sudo.partner_id.id
+
+    def _find_child_partner(self, commercial_partner_id, address):
+        """ Find a child partner for a specified address
+
+        Compare all keys in the `address` dict with the same keys on the partner object and return
+        the id of the first partner that have the same value than in the dict for all the keys.
+
+        :param int commercial_partner_id: commercial partner for whom we need to find his children.
+        :param dict address: dictionary of address fields.
+        :return int: id of the first child partner that match the criteria, if any.
+        """
+        partners_sudo = request.env['res.partner'].with_context(
+            show_address=1
+        ).sudo().search([
+            ('id', 'child_of', commercial_partner_id),
+        ])
+        for partner_sudo in partners_sudo:
+            if all(address[k] == partner_sudo[k] for k in address):
+                return partner_sudo.id
+        return False
+
+    def _include_country_and_state_in_address(self, address):
+        """ This function is used to include country_id and state_id in address.
+
+        Fetch country and state and include the records in address. The object is included to
+        simplify the comparison of addresses.
+
+        :param dict address: An address with country and state defined in ISO 3166.
+        :return None:
+        """
+        country = request.env["res.country"].search([
+            ('code', '=', address.pop('country')),
+        ], limit=1)
+        state = request.env["res.country.state"].search([
+            ('code', '=', address.pop('state')),
+        ], limit=1)
+        address['country_id'] = country
+        address['state_id'] = state
```

Comment:
```suggestion
        address.update(country_id=country,state_id=state)
```
Prefer using kwargs rather than call `dict.update` with a dict literal :) Good suggestion though ^^

Reply:
Nice!! Thanks, I'll keep that in mind! =D